packages/std/src/math.rs

```diff
@@ -65,42 +65,36 @@ impl FromStr for Decimal {
     /// This never performs any kind of rounding.
     /// More than 18 fractional digits, even zeros, result in an error.
     fn from_str(input: &str) -> Result<Self, Self::Err> {
-        let parts: Vec<&str> = input.split('.').collect();
-        match parts.len() {
-            1 => {
-                let whole = parts[0]
-                    .parse::<u128>()
-                    .map_err(|_| StdError::generic_err("Error parsing whole"))?;
-
-                let whole_as_atomics = whole
-                    .checked_mul(DECIMAL_FRACTIONAL)
-                    .ok_or_else(|| StdError::generic_err("Value too big"))?;
-                Ok(Decimal(whole_as_atomics))
-            }
-            2 => {
-                let whole = parts[0]
-                    .parse::<u128>()
-                    .map_err(|_| StdError::generic_err("Error parsing whole"))?;
-                let fractional = parts[1]
-                    .parse::<u128>()
-                    .map_err(|_| StdError::generic_err("Error parsing fractional"))?;
-                let exp = (18usize.checked_sub(parts[1].len())).ok_or_else(|| {
-                    StdError::generic_err("Cannot parse more than 18 fractional digits")
-                })?;
-                let fractional_factor = 10u128
-                    .checked_pow(exp.try_into().unwrap())
-                    .ok_or_else(|| StdError::generic_err("Cannot compute fractional factor"))?;
-
-                let whole_as_atomics = whole
-                    .checked_mul(DECIMAL_FRACTIONAL)
-                    .ok_or_else(|| StdError::generic_err("Value too big"))?;
-                let atomics = whole_as_atomics
-                    .checked_add(fractional * fractional_factor)
-                    .ok_or_else(|| StdError::generic_err("Value too big"))?;
-                Ok(Decimal(atomics))
-            }
-            _ => Err(StdError::generic_err("Unexpected number of dots")),
+        let mut parts_iter = input.split('.');
+
+        let whole_part = parts_iter.next().unwrap(); // split always returns at least one element
+        let whole = whole_part
+            .parse::<u128>()
+            .map_err(|_| StdError::generic_err("Error parsing whole"))?;
+        let mut atomics = whole
+            .checked_mul(DECIMAL_FRACTIONAL)
+            .ok_or_else(|| StdError::generic_err("Value too big"))?;
+
+        if let Some(fractional_part) = parts_iter.next() {
+            let fractional = fractional_part
+                .parse::<u128>()
+                .map_err(|_| StdError::generic_err("Error parsing fractional"))?;
+            let exp = (18usize.checked_sub(fractional_part.len())).ok_or_else(|| {
+                StdError::generic_err("Cannot parse more than 18 fractional digits")
+            })?;
+            let fractional_factor = 10u128
+                .checked_pow(exp.try_into().unwrap())
+                .ok_or_else(|| StdError::generic_err("Cannot compute fractional factor"))?;
```

Comment:
```suggestion
            let exp = (18u32.checked_sub(fractional_part.len() as u32)).ok_or_else(|| {
                StdError::generic_err("Cannot parse more than 18 fractional digits")
            })?;
            let fractional_factor = 10u128.pow(exp);
```
That `pow` cannot fail, right?

Reply:
I think you are right. Will add this change manually with a comment or two.

Reply:
Yes, better to add it manually, as it changes some imports above.

Reply:
Here we go: https://github.com/CosmWasm/cosmwasm/pull/721/commits/9cb8fb4a0592a6be6f09a006f1caa991a8e1de46